### ISSUE TITLE
group_send getting delayed while sending to active window when a user has multiple windows open

### ISSUE THREAD

**Issue by niranjanrao1** [django/channels#1730]
Background: I'm working on a chat application on my local (development) server. Python 3.9.5, Django 3.2, Channels 3.0.3. Each user can see an "all messages" section on the page, containing messages from all users, and a "my messages" section, containing only his own messages.

Goal: All messages should be sent to all users, and each user's messages should be separately sent only to him/her. If the user is logged in via multiple browser tabs/ devices (this scenario is supposed to be handled by Channels as per the docs [here](https://channels.readthedocs.io/en/stable/topics/channel_layers.html#groups)), and the user sends a message from one window, it needs to immediately send back (echo) the message to all his windows, including the window where the message was sent from.

For this, I'm adding each connection to a group when it is first opened.

Consumers.py

    async def websocket_connect(self, event):
        await self.send({
            "type": "websocket.accept"
        })
        await self.channel_layer.group_add("master", self.channel_name)
        user = self.scope['user']
        user_group = "user_" + str(user.id)
        if user.is_authenticated:
            await self.channel_layer.group_add(user_group, self.channel_name)        

As you can see, the master group is for all users, and the user_group is for all connections of this particular user.

Further:

    async def websocket_receive(self, event):
        user = self.scope['user']
        user_group = "user_" + str(user.id)
        ....some code...

        await self.channel_layer.group_send(
            "master",{
                'type': 'master.broadcast',
                'text': json.dumps(messages_all)
            }
        )
        await self.channel_layer.group_send(
            user_group,{
                'type': 'user.broadcast',
                'text': json.dumps(messages_user)
            }
        )

The broadcast functions are simple send functions:

    async def master_broadcast(self, event):
        await self.send({
            "type": "websocket.send",
            "text": event['text']
        })
    
    async def user_broadcast(self, event):
        await self.send({
            "type": "websocket.send",
            "text": event['text']
        })

The problem:

When the user sends a message from one of his windows, all connections in the "master" group (from the first group_send()) are successfully getting the messages, except the user's own active window from which the message was sent. Here, there's a delay of about 10-20 seconds, after which the user's active window receives the master message. Immediately, all the other messages (to user_group, via the 2nd group_send()) are sent without any more delay (including the active window).

Why is this happening and how can I fix it?

Grateful for any help here. To be clear, if the user is logged in from only 1 window/tab, there's no problem and the messages are delivered instantly.

P.S: I have already tried sending messages from each of the different windows of the user. In each case, the delay is only in the case of the window that is sending the message. SO this has nothing to do with the browsers being used etc.

Below are the answers to the other notes suggested here:

- Your OS and runtime environment, and browser if applicable - Windows 10 Enterprise Edition, python 3.9.5, browsers Firefox, Chrome and IE simultaneously
- A `pip freeze` output showing your package versions - this would be too long, but I'm using Django 3.2 and Channels 3.0.3
- What you expected to happen vs. what actually happened: I expect all the browser windows to get the message instantly (since they're part of the same "group"). They do get the message, but with a lag in-between. Kindly read the full description below for the full sequence of events
- How you're running Channels (runserver? daphne/runworker? Nginx/Apache in front?): runserver
- Console logs and full tracebacks of any errors; There is no error, just a massive delay when a user is logged in on multiple windows/ consumers (so the user's group size is more than 1).

**Comment by richardARPANET** [comment]
I'm seeing the following behavior myself, with this setup.

JS client. Python server. Client sends to server, it hits AsyncJsonWebsocketConsumer, that does `self.channel_layer.send('background-tasks', content)`

then background is handled by a `AsyncConsumer` which sends back `self.channel_layer.group_send(group_name, message)`

this is then forwarded by the original AsyncJsonWebsocketConsumer to the client.

1. Start 1 browser window, 2 way websockets working.
2. Open another browser tab, client is able to sent but not recieve anything, they never reach the background `AsyncConsumer`.

**Comment by brycedrennan** [comment]
Just ran into this exact issue. Trying to do the same thing. 

Python 3.12, 
Django==5.1.1
channels==4.1.0
channels-redis==4.2.0

**Comment by brycedrennan** [comment]
The cause is that the thread handling the messages for the sending tab is blocked by receiving the message. I solved by putting the work into a asyncio task and thus allowing the while True loop to process the messages that are being sent by the task

```python

    async def disconnect(self, close_code):
        await self.channel_layer.group_discard(
            self.conversation_group_name, self.channel_name
        )

    async def receive(self, text_data):
        data = json.loads(text_data)
        # Start a new task to process the message
        asyncio.create_task(self.process_message(data))
   ```
